### DISCUSSION TITLE
Would there ever be a case where we would have to limit the allowed db types of a mathesar type on the client?

### DISCUSSION THREAD

**Discussion by pavish** [mathesar-foundation/mathesar#935] [Developer Q&A] [answered]
The column endpoint currently returns a property `valid_target_types` which is a list of allowed target database types.

However on the frontend, we do not allow the user to directly choose the database type. Instead, we display mathesar types and identify the database type based on the configurations entered by the user.

Would there be a case where the specified configuration might lead to the selection of a database type which is not an allowed target?

For eg.,
Assume the values returned by `valid_target_types` only contains `['INTEGER', 'BIGINT', 'SMALLINT', 'NUMERIC']`.

On the frontend however, based on our UX, we display a configuration pane:

<img width="268" alt="Screenshot 2022-01-03 at 12 13 51 PM" src="https://user-images.githubusercontent.com/11850603/147905073-b9183b89-4fa7-4c12-8c65-0a8fa09bf9ca.png">

The user is free to choose whatever configuration here, leading to the selected db type being '`DECIMAL`'. This is not an allowed target type.

Is this scenario possible?

Additional questions:
* If this scenario should not be possible, should valid_target_types send mathesar types instead of database types?
* If this scenario is possible, what changes would our current UX require? How can we construct a type agnostic UI for it?

**Comment by mathemancer** [comment, marked as answer]
If something isn't listed by `valid_target_types`, and the front end is funneling users into casting to that type, that's a bug.  We _cannot_ cast a column to any DB type not listed in `valid_target_types`, as this lists every target type for which a casting function is implemented.  Thus, the scenario should not be possible, and if it occurs, we should submit a bug and fix it.

A scenario that could theoretically happen would be where only some subset of the Mathesar type is available as a target for a given type.  I can't see any that makes logical sense with our current types, but it's feasible.  For a contrived example, suppose have both emails and URLs under some "address" type in the future, and I'm coming from a URI.  I can't cast any valid URI to an email, but I can cast some valid URIs to URLs.

As for sending the `mathesar_types` rather than DB types, I'd rather keep sending the DB type, since it would allow us to notice and handle the bug of missing some desired target DB type earlier and handle it more gracefully.  It also gives more information and power to the front end, and the mapping between types is available via another endpoint for the front end to check whether it's choosing something valid if desired.  I also like having the DB level info available, since there may be some situations where another client of the API has use for the info.

I'm not fiercely opposed to changing this behavior, however, it's a mild preference.

**Reply by kgodey** [reply]
I'm going to mark this response as the answer, I think it addresses everything in the original question.
